homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,251 @@
+"""
+Support for a generic MQTT vacuum.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/vacuum/
+"""
+import asyncio
+import json
+import logging
+
+import voluptuous as vol
+
+import homeassistant.components.mqtt as mqtt
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.vacuum import (
+    DEFAULT_ICON, SUPPORT_BATTERY, SUPPORT_CLEAN_SPOT,
+    SUPPORT_LOCATE, SUPPORT_PAUSE, SUPPORT_RETURN_HOME,
+    SUPPORT_STATUS, SUPPORT_STOP, SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON, VacuumDevice, services_to_strings, strings_to_services)
+from homeassistant.const import CONF_NAME, ATTR_SUPPORTED_FEATURES,\
+    ATTR_BATTERY_LEVEL, ATTR_STATE
+from homeassistant.core import callback
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['mqtt']
+
+DEFAULT_SERVICES = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_STOP |\
+                   SUPPORT_RETURN_HOME | SUPPORT_STATUS | SUPPORT_BATTERY |\
+                   SUPPORT_CLEAN_SPOT
+ALL_SERVICES = DEFAULT_SERVICES | SUPPORT_PAUSE | SUPPORT_LOCATE
+DEFAULT_SERVICE_STRINGS = services_to_strings(DEFAULT_SERVICES)
+
+DEFAULT_COMMAND_TOPIC = 'vacuum/command'
+DEFAULT_STATE_TOPIC = 'vacuum/state'
+
+DEFAULT_NAME = 'MQTT Vacuum'
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend({
+    vol.Optional(ATTR_SUPPORTED_FEATURES, default=DEFAULT_SERVICE_STRINGS):
+        vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(mqtt.CONF_COMMAND_TOPIC, default=DEFAULT_COMMAND_TOPIC):
+        mqtt.valid_publish_topic,
+    vol.Optional(mqtt.CONF_STATE_TOPIC, default=DEFAULT_STATE_TOPIC):
+        mqtt.valid_publish_topic,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```

Comment:
Use the async version `async_setup_platform`. See eg https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/mqtt.py.

Reply:
Will do. Thanks for the link!